repo/blob/azure/azure_storage_test.go

```diff
@@ -270,3 +273,125 @@ func TestAzureStorageInvalidCreds(t *testing.T) {
 		t.Errorf("unexpected success connecting to Azure blob storage, wanted error")
 	}
 }
+
+// TestAzureStorageImmutabilityProtection runs through the behavior of Azure immutability protection.
+func TestAzureStorageImmutabilityProtection(t *testing.T) {
+	t.Parallel()
+	testutil.ProviderTest(t)
+
+	// must be with ImmutableStorage with Versioning enabled
+	container := getEnvOrSkip(t, testContainerEnv)
+	storageAccount := getEnvOrSkip(t, testStorageAccountEnv)
+	storageKey := getEnvOrSkip(t, testStorageKeyEnv)
+
+	// create container if does not exist
+	createContainer(t, container, storageAccount, storageKey)
+
+	data := make([]byte, 8)
+	rand.Read(data)
+
+	ctx := testlogging.Context(t)
+
+	// use context that gets canceled after opening storage to ensure it's not used beyond New().
+	newctx, cancel := context.WithCancel(ctx)
+	prefix := fmt.Sprintf("test-%v-%x-", clock.Now().Unix(), data)
+	st, err := azure.New(newctx, &azure.Options{
+		Container:      container,
+		StorageAccount: storageAccount,
+		StorageKey:     storageKey,
+		Prefix:         prefix,
+	}, false)
+
+	cancel()
+	require.NoError(t, err)
+
+	defer st.Close(ctx)
+
+	const (
+		blobName  = "sExample"
+		dummyBlob = blob.ID(blobName)
+	)
+
+	blobNameFullPath := prefix + blobName
+
+	putOpts := blob.PutOptions{
+		RetentionMode:   blob.Locked,
+		RetentionPeriod: 3 * time.Second,
+	}
+	// non-nil blob to distinguish against delete marker version
+	err = st.PutBlob(ctx, dummyBlob, gather.FromSlice([]byte("x")), putOpts)
+	require.NoError(t, err)
+	cli := getAzureCLI(t, storageAccount, storageKey)
+
+	count := getBlobCount(ctx, t, st, content.BlobIDPrefixSession)
+	require.Equal(t, 1, count)
+
+	currentTime := clock.Now().UTC()
+
+	blobRetention := getBlobRetention(ctx, t, cli, container, blobNameFullPath)
+	if !blobRetention.After(currentTime) {
+		t.Fatalf("blob retention period not in the future: %v", blobRetention)
+		t.FailNow()
+	}
+
+	extendOpts := blob.ExtendOptions{
+		RetentionMode:   blob.Locked,
+		RetentionPeriod: 4 * time.Second,
+	}
+	err = st.ExtendBlobRetention(ctx, dummyBlob, extendOpts)
```

Comment:
This will not pass until https://github.com/kopia/kopia/pull/3412 is merged

Reply:
It will also not pass until https://learn.microsoft.com/en-us/azure/storage/blobs/immutable-policy-configure-version-scope?tabs=azure-portal#enable-support-for-version-level-immutability is enabled - would you be able to do this for the relevant provider @jkowalski?

Reply:
Correct. @julio-lopez said he is working on a script from this.

Reply:
Container has been set up, and has been used for the Azure PITR functionality.